src/Exceptions/ExceptionHandler.php

```
@@ -41,6 +68,7 @@ public function report(Exception $e)
         }
 
         $this->logger->error($e->getMessage(), $context);
+        $this->appExceptionHandler->report($e);
```

I'm not sure about that line.

Basically, we're doing double logging here:

```php
// Log into the command's folder (package's main intention)
$this->logger->error($e->getMessage(), $context);

// Log into the "standard" Laravel log file (wasn't that way before)
$this->appExceptionHandler->report($e);
```

---

In that context, this class is not really a decorator for the Laravel's exceptions `Handler`. It overwrites functionality of the `report()` method, NOT decorates it. However, I do understand that it also means that the Sentry's decorator potentially would NOT be called (it depends on the order of "decorations").

Initially, I thought that we could keep this call to the parent class, but simply return `false` from `shouldReport()` method. But I think it would NOT help us too, because when the call would be passed to the parent class, it will use its own "protected" method and do the logging anyways. And also, it might cause other problems if we return "false" for the "outer world" from our decorated exceptions `Handler`...

Do you have any ideas about that?

---

Hi, @dmitry-ivanov . Thank you for interesting comments! 

> I'm not sure about that line.

I think this is a essence of the "Decorator "pattern. We need use behavior of previous object and add our behavior.

> Basically, we're doing double logging here:

Yes, but for `Exceptions` only. I don't see the problem if we will have error messages in common log and command log.

> In that context, this class is not really a decorator for the Laravel's exceptions Handler. It overwrites functionality of the report() method, NOT decorates it. 

If we need have functionality of another `ExceptionHandler`, I don't see another way than Decorator still.

---

Or we can do as Sentry do. It does not make own ExceptionHandler.

https://sentry.io/for/laravel/  :

> Add Sentry reporting to App/Exceptions/Handler.php:

```
public function report(Exception $exception)
{
    if (app()->bound('sentry') && $this->shouldReport($exception)) {
        app('sentry')->captureException($exception);
    }

    parent::report($exception);
}
```

Also we can add something like this:

```
public function report(Exception $exception)
{
    if (app()->bound('sentry') && $this->shouldReport($exception)) {
        app('sentry')->captureException($exception);
    }


    parent::report($exception);
}
```

---

@itelmenko Thanks for continuing helping on that! 👍 

Good to know about that approach which Sentry uses!

I definitely don't want to complicate the package installation (I mean requiring that additional step for adding some code in `App\Exceptions\Handler`), but at the same time, I have an idea.

What if we add that in the package's `Illuminated\Console\Exceptions\ExceptionHandler`:

```php
public function report(Exception $e)
{
   // So in the package itself, we'd add that integration with Sentry 
   if (app()->bound('sentry') && $this->shouldReport($exception)) {
        app('sentry')->captureException($exception);
    }
    
    // Everything below - is just the current body of "report()" method
    $context = [
        'code' => $e->getCode(),
        'message' => $e->getMessage(),
        'file' => $e->getFile(),
        'line' => $e->getLine(),
    ];

    if ($e instanceof RuntimeException) {
        $eContext = $e->getContext();
        if (!empty($eContext)) {
            $context['context'] = $eContext;
        }
    }

    $this->logger->error($e->getMessage(), $context);
}
```

Of course, the downside is that we'd have to do the same for other services like Sentry, but it's a tradeoff, and the benefit is that we wouldn't change the package installation process.

And also I haven't received any requests about other services except Sentry, so I'm totally ok with that.

How do you feel about that?

---

Please, lets take a look at Decorator patter again.

**Advantages**
1. The Decorator doesn't require extra step for installation
2. The Decorator will work fine with another services (not only Sentry)
3. The Decorator will not redefine current ExceptionHandler  -`App\Exceptions\ExceptionHandler`. Application can have another functionality in its ExceptionHanldler. We will not get side effects in the application. The end user does not expect the ExceptionHandler to be overridden.

**Disadvantages** 

1. We will have error notification in two places: in common log and command's log. But I am not sure it is a disadvantage.

With all due respect, in the same time, the snippet

```php
// So in the package itself, we'd add that integration with Sentry 
   if (app()->bound('sentry') && $this->shouldReport($exception)) {
        app('sentry')->captureException($exception);
    }
```

in `Illuminated\Console\Exceptions\ExceptionHandler` is hard code and looks like dirty hack.

So I think approach with Decorator would be better...

---

About this single disadvantage of approach with Decorator. User can prevent sending command's exceptions to common log.
He can implement `render()` method of  `App\Exceptions\ExceptionHandler` like this:

```php
    public function report(Exception $exception)
    {
        if (app()->bound('sentry') && $this->shouldReport($exception)) {
            // Sending to Sentry
            app('sentry')->captureException($exception);
        }
        if(strpos($exception->getFile(), '/app/Console/Commands/') !== false) {
            return;
        }
        parent::report($exception);
    }
```

---

@itelmenko Thank you for taking the time to work on that! I really appreciate your help! 👍 

**Advantages**

> The Decorator doesn't require extra step for installation

Agree 👍 

> The Decorator will work fine with another services (not only Sentry)

Agree 👍 

> The Decorator will not redefine current ExceptionHandler -App\Exceptions\ExceptionHandler. Application can have another functionality in its ExceptionHanldler. We will not get side effects in the application. The end user does not expect the ExceptionHandler to be overridden.

I see your point, but actually that SHOULD be expected. The main idea of the package is to "redirect" all output from the command, and accumulate it in a separate place. That's the core idea of the package. In that sense, we DO change the behavior of the application's exception handler, and we're NOT decorating it. The user explicitly says that he wants that changed behavior by using `Loggable` trait in his command. So, the package would NOT overwrite the app's exception handler for all cases, it happens only when user explicitly says that he wants THAT.

At the same time, I DO see your point, and I feel your pain with Sentry (and potentially other custom logic defined in app's exception handler).

I agree that this solution with putting the "Sentry logic" into the package is a "hack", I'm not worried about that too much, because Sentry integration was the only one question related to that topic so far.

However, we might try to make it more flexible. How do you feel about idea of putting there some callback instead, so that the user would be able to inject any additional logic for "report()" method?

Something like that:

```php
public function report(Exception $e)
{
    // Some callback here, in which "Sentry" and any other "custom logic" could go

    // Current body of "report()"...
}
```

I'm just not sure how exactly to provide that callback yet, but we might figure it out.

---

There are more thing than just the Sentry (in `App\Exceptions\ExceptionHandler`).

For example:

```php

/**
     * A list of the exception types that are not reported.
     *
     * @var array
     */
    protected $dontReport = [
        //
    ];

    /**
     * A list of the inputs that are never flashed for validation exceptions.
     *
     * @var array
     */
    protected $dontFlash = [
        'password',
        'password_confirmation',
    ];

```

It is default code. But in real application there can be something more. Also we may miss other features of application's `ExceptionHandler`. I think it is not good idea to replace default `ExceptionHandler` instead of extending.

>  I'm not worried about that too much, because Sentry integration was the only one question related to that topic so far.

I think this question (about default `ExceptionHandler`) will appear more often if the package will become more popular. If you need change something it will be better change it earlier. Fewer users will experience inconvenience.

---

> There are more things than just the Sentry (in App\Exceptions\ExceptionHandler).

Yeah, but at the same time - both of these things (`$dontReport` and `$dontFlash`) are relevant only to HTTP requests. I never needed to ignore any exceptions in my console commands, and if someone will ever ask for it - we can easily add it. The framework is pretty stable now, so I don't believe things would change drastically in the parent exception handler.

I totally see your point, but that's my intentional choice - this package DOES change the behavior of the app's exception handler. I understand that it might be not applicable to all cases in the world, but I'm fine with that.

I'll add Sentry support in the separate PR.